lib/db_adapter.go

```diff
@@ -62,32 +65,132 @@ func (adapter *DbAdapter) GetPostAssociationByAttributes(associationEntry *PostA
-		return adapter.postgresDb.GetUserAssociationsByAttributes(associationQuery)
+		return adapter.postgresDb.GetUserAssociationsByAttributes(associationQuery, deletedUtxoAssociationIDs)
 	}
-	return DBGetUserAssociationsByAttributes(adapter.badgerDb, adapter.snapshot, associationQuery)
+	return DBGetUserAssociationsByAttributes(adapter.badgerDb, adapter.snapshot, associationQuery, deletedUtxoAssociationIDs)
 }
 
-func (adapter *DbAdapter) GetPostAssociationsByAttributes(associationQuery *PostAssociationQuery) ([]*PostAssociationEntry, error) {
+func (adapter *DbAdapter) GetPostAssociationsByAttributes(
+	associationQuery *PostAssociationQuery,
+	deletedUtxoAssociationIDs *Set[*BlockHash],
+) ([]*PostAssociationEntry, []byte, error) {
 	if adapter.postgresDb != nil {
-		return adapter.postgresDb.GetPostAssociationsByAttributes(associationQuery)
+		return adapter.postgresDb.GetPostAssociationsByAttributes(associationQuery, deletedUtxoAssociationIDs)
 	}
-	return DBGetPostAssociationsByAttributes(adapter.badgerDb, adapter.snapshot, associationQuery)
+	return DBGetPostAssociationsByAttributes(adapter.badgerDb, adapter.snapshot, associationQuery, deletedUtxoAssociationIDs)
 }
 
-func (adapter *DbAdapter) GetUserAssociationIdsByAttributes(associationQuery *UserAssociationQuery) ([]*BlockHash, error) {
+func (adapter *DbAdapter) GetUserAssociationIdsByAttributes(
+	associationQuery *UserAssociationQuery, deletedUtxoAssociationIDs *Set[*BlockHash],
+) (*Set[*BlockHash], []byte, error) {
 	if adapter.postgresDb != nil {
-		return adapter.postgresDb.GetUserAssociationIdsByAttributes(associationQuery)
+		return adapter.postgresDb.GetUserAssociationIdsByAttributes(associationQuery, deletedUtxoAssociationIDs)
 	}
-	return DBGetUserAssociationIdsByAttributes(adapter.badgerDb, adapter.snapshot, associationQuery)
+	return DBGetUserAssociationIdsByAttributes(adapter.badgerDb, adapter.snapshot, associationQuery, deletedUtxoAssociationIDs)
 }
 
-func (adapter *DbAdapter) GetPostAssociationIdsByAttributes(associationQuery *PostAssociationQuery) ([]*BlockHash, error) {
+func (adapter *DbAdapter) GetPostAssociationIdsByAttributes(
+	associationQuery *PostAssociationQuery, deletedUtxoAssociationIDs *Set[*BlockHash],
+) (*Set[*BlockHash], []byte, error) {
 	if adapter.postgresDb != nil {
-		return adapter.postgresDb.GetPostAssociationIdsByAttributes(associationQuery)
+		return adapter.postgresDb.GetPostAssociationIdsByAttributes(associationQuery, deletedUtxoAssociationIDs)
 	}
-	return DBGetPostAssociationIdsByAttributes(adapter.badgerDb, adapter.snapshot, associationQuery)
+	return DBGetPostAssociationIdsByAttributes(adapter.badgerDb, adapter.snapshot, associationQuery, deletedUtxoAssociationIDs)
+}
+
+func (adapter *DbAdapter) SortUserAssociationEntriesByPrefix(
+	associationEntries []*UserAssociationEntry,
+	prefixType []byte,
+	sortDescending bool,
+) ([]*UserAssociationEntry, error) {
+	// Postgres sorts results by AssociationID.
+	if adapter.postgresDb != nil {
```

Comment:
DBKeyForEntry is in-memory whereas postgres sorts by AssociationID in the database. IMO it's ok for postgres and badger to diverge in the non-consensus querying. Postgres has more lax rules around parameters (since order of badger key doesn't matter). And paginates through association entries by AssociationID instead of badger key prefix.

Reply:
sounds good - resolved